components/apimgt/org.wso2.carbon.apimgt.jms.listener/src/main/java/org/wso2/carbon/apimgt/jms/listener/utils/JMSMessageListener.java

```diff
@@ -124,6 +149,54 @@ public void onMessage(Message message) {
         }
     }
 
+    private void handleUserCacheInvalidationMessage(Map<String, Object> map) throws ParseException {
+
+        if (map.containsKey("value")) {
+            String value = (String) map.get("value");
+            JSONParser jsonParser = new JSONParser();
+            JSONArray jsonValue = (JSONArray) jsonParser.parse(value);
+
+            ServiceReferenceHolder.getInstance().getCacheInvalidationService().invalidateCachedUsernames((String[]) jsonValue.toArray(new String[jsonValue.size()]));
+        }
+    }
+
+    private void handleKeyCacheInvalidationMessage(Map<String, Object> map) throws ParseException {
+
+        if (map.containsKey("value")) {
+            String value = (String) map.get("value");
+            JSONParser jsonParser = new JSONParser();
+            JSONArray jsonValue = (JSONArray) jsonParser.parse(value);
+
+            ServiceReferenceHolder.getInstance().getCacheInvalidationService().invalidateCachedTokens((String[]) jsonValue.toArray(new String[jsonValue.size()]));
+        }
+    }
+
+    private void handleResourceCacheInvalidationMessage(Map<String, Object> map) throws ParseException {
+
+        if (map.containsKey("value")) {
+            String value = (String) map.get("value");
+            JSONParser jsonParser = new JSONParser();
+            JSONObject jsonValue = (JSONObject) jsonParser.parse(value);
+            String apiContext = (String) jsonValue.get("apiContext");
+            String apiVersion = (String) jsonValue.get("apiVersion");
+            JSONArray resources = (JSONArray) jsonValue.get("resources");
+            List<ResourceCacheInvalidationDto> resourceCacheInvalidationDtoList = new ArrayList<>();
+            for (Object resource : resources) {
+                JSONObject uriTemplate = (JSONObject) resource;
+                String resourceURLContext = (String) uriTemplate.get("resourceURLContext");
+                String httpVerb = (String) uriTemplate.get("httpVerb");
+                ResourceCacheInvalidationDto uriTemplateDto = new ResourceCacheInvalidationDto();
+                uriTemplateDto.setHttpVerb(httpVerb);
+                uriTemplateDto.setResourceURLContext(resourceURLContext);
+                resourceCacheInvalidationDtoList.add(uriTemplateDto);
+            }
+            ServiceReferenceHolder.getInstance().getCacheInvalidationService().invalidateResourceCache(apiContext,
+                    apiVersion,
+                    resourceCacheInvalidationDtoList.toArray(new ResourceCacheInvalidationDto[resourceCacheInvalidationDtoList.size()]));
```

Comment:
here too

Reply:
We need set of debug loes which resources cleared and whether we recenved the invalifdation msf and etc for analyzing issues